crates/bevy_ui/src/lib.rs

```diff
@@ -178,72 +200,95 @@ impl Plugin for UiPlugin {
-                    // the systems don't care about stack index
-                    .ambiguous_with(update_clipping_system)
-                    .ambiguous_with(ui_layout_system)
-                    .in_set(AmbiguousWithTextSystem),
-                update_clipping_system.after(TransformSystem::TransformPropagate),
-                // Potential conflicts: `Assets<Image>`
-                // They run independently since `widget::image_node_system` will only ever observe
-                // its own ImageNode, and `widget::text_system` & `bevy_text::update_text2d_layout`
-                // will never modify a pre-existing `Image` asset.
-                widget::update_image_content_size_system
-                    .in_set(UiSystem::Prepare)
-                    .in_set(AmbiguousWithTextSystem)
-                    .in_set(AmbiguousWithUpdateText2DLayout),
             ),
         );
+
+        #[cfg(feature = "bevy_render")]
+        {
+            let systems = ui_layout_system
+                .in_set(UiSystem::Layout)
+                .before(TransformSystem::TransformPropagate);
+            #[cfg(feature = "bevy_text")]
+            let systems = systems
+                // Text and Text2D operate on disjoint sets of entities
+                .ambiguous_with(bevy_text::update_text2d_layout)
+                .ambiguous_with(bevy_text::detect_text_needs_rerender::<bevy_text::Text2d>);
+            app.add_systems(PostUpdate, systems);
+        }
+
+        let systems = ui_stack_system
+            .in_set(UiSystem::Stack)
+            // the systems don't care about stack index
+            .ambiguous_with(update_clipping_system);
+        #[cfg(feature = "bevy_render")]
+        let systems = systems.ambiguous_with(ui_layout_system);
+        #[cfg(feature = "bevy_text")]
+        let systems = systems.in_set(AmbiguousWithTextSystem);
+        app.add_systems(PostUpdate, systems);
+        app.add_systems(
+            PostUpdate,
+            update_clipping_system.after(TransformSystem::TransformPropagate),
+        );
+
+        // Potential conflicts: `Assets<Image>`
+        // They run independently since `widget::image_node_system` will only ever observe
+        // its own ImageNode, and `widget::text_system` & `update_text2d_layout`
+        // will never modify a pre-existing `Image` asset.
+        let systems = widget::update_image_content_size_system.in_set(UiSystem::Prepare);
+        #[cfg(feature = "bevy_text")]
+        let systems = systems
+            .in_set(AmbiguousWithTextSystem)
+            .in_set(AmbiguousWithUpdateText2DLayout);
+        app.add_systems(PostUpdate, systems);
+
+        #[cfg(feature = "bevy_text")]
         build_text_interop(app);
 
         #[cfg(feature = "bevy_ui_picking_backend")]
         app.add_plugins(picking_backend::UiPickingPlugin);
 
+        #[cfg(feature = "bevy_render")]
         if !self.enable_rendering {
             return;
         }
```

Comment:
```suggestion
```

I don't think there is a need for both this and the feature gate.

Reply:
It would be dead code